### PR TITLE
Topic recognition

### PR DESCRIPTION
**Pull request by Shriyan-Garg**
Dear COMP3710 team,
For this project, I have done task 6 where we need to use the ADNI brain dataset and build a Vision Transformer (ViT) to classify whether the brain scan shows evidence of Alzheimer’s disease (AD) or is it normal (NC). My implementation contains 5 files:

`dataset.py:` Defines a custom dataset class for AD and NC image data, along with associated transformations and data loaders for training and testing.

`modules.py:` This module defines the neural network components utilized in the Vision Transformer architecture. It provides classes for patch embedding and the main Vision Transformer model suitable for image classification tasks.

`train.py:` Initializes, trains, and test the VisionTransformer on AD and NC image data. After training, the model is saved as well. It also provides visualization of training and testing metrics.

`predict.py:` This script loads a trained VisionTransformer model and uses it to predict the class of a given image. It then visualizes the input image and provides a confidence score bar chart.

`README.md:` This file contains a detailed report about the project.

**General reflection about the project**: The project was quite interesting as it allowed me to learn about computer vision. My model is currently able to get 65.5% accuracy on the test set. One of the reasons for low accuracy was that I used Google Colab, which has certain limitations on using the T4 GPU. The maximum time we could use T4 GPU on Google Colab is around 3 hours, after that, the session disconnects. Because of this, I need to limit the number of epochs. It is also important to note that several changes could be made in the code for better performance.
Overall, it was a good experience working on a project that could help in solving a real-life problem. Looking forward to the staff’s valuable feedback.

Regards
Shriyan Garg
46760807 (s4676080)

### CODE REVIEW

**Comment by LinfengLiu98** [comment]
**This is an initial inspection, no action is required**

**Difficulty**: 
- Hard
- Task 6: . Classify Alzheimer’s disease
- Vision transformer

**Readme**
- Missing Vision transformer algorithm explanation

**Commit messages**
- Good

**Code**
- Good structure
- Adequate comments

**Functionality/Performance**
- 66.5% classification accuracy
- Did patient level split

**General comments**
- Good report
- Good plots
- No changes required

**Comment by Shriyan-Garg** [comment]
Thanks for the review and your feedback

Kindly let me know if I can update my README file by adding the explanation of Vision Transformer Algorithm.

Regards
Shriyan Garg

**Comment by Shriyan-Garg** [comment]
Greetings,
This is to inform you that I have incorporated the feedback by adding a detailed explanation of ViT algorithm in README.

Thanks
Shriyan Garg
46760807

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks
## Recognition Problem 
Solves problem, but the performance is < 60%, could use more training? -1
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable, missing usage info -1
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
